infiniband-diags/ibportstate.c

```diff
@@ -402,7 +402,7 @@ int main(int argc, char **argv)
 	    "\tmkey, mkeylease, mkeyprot\n";
 	const char *usage_examples[] = {
 		"3 1 disable\t\t\t# by lid",
-		"-G 0x2C9000100D051 1 enable\t# by guid",
+		"3 1 -C mlx4_0 -P 1 enable\t# by lid, CA name and Port number",
```

Comment:
Since when is enablement via GUID not working?

FWIW if a CA port is disabled it does not have a valid LID (unless running on an SM node).  So I'm wondering why this is better than using a port GUID which should be unique.

Reply:
[root ~]$ ibstat
CA 'mlx4_0'
	CA type: MT4099
	Number of ports: 2
	Firmware version: 2.42.5000
	Hardware version: 1
	Node GUID: 0xf4521403007be0e0
	System image GUID: 0xf4521403007be0e3
	Port 1:
		State: Down
		Physical state: Disabled
		Rate: 10
		Base lid: 38
		LMC: 0
		SM lid: 13
		Capability mask: 0x02594868
		Port GUID: 0xf4521403007be0e1
		Link layer: InfiniBand
	Port 2:
		State: Active
		Physical state: LinkUp
		Rate: 56
		Base lid: 3
		LMC: 0
		SM lid: 1
		Capability mask: 0x02594868
		Port GUID: 0xf4521403007be0e2
		Link layer: InfiniBand


[root ~]$ ibportstate -G 0xf4521403007be0e1 1 enable
ibwarn: [84425] ib_path_query_via: sa call path_query failed
ibportstate: iberror: failed: can't resolve destination port 0xf4521403007be0e1

Reply:
> Since when is enablement via GUID not working?

I did not run the git bisect test. But I build ibportstate with code from upstream repo. The problem is that the first port is disabled. function 'resolve_ca_port' skips the first port.

```c
 292├>        for (i = 0; i <= ca.numports; i++) {
 293│                 DEBUG("checking port %d", i);
 294│                 if (!ca.ports[i])
 295│                         continue;
 296│                 if (strcmp(ca.ports[i]->link_layer, "InfiniBand") &&
 297│                     strcmp(ca.ports[i]->link_layer, "IB"))
 298│                         continue;
 299│                 if (up < 0 && ca.ports[i]->phys_state == 5)
 300│                         up = *port = i;
/root/rdma-core/libibumad/umad.c  
```                                                                                                                

(gdb) p	ca.ports[0]
$11 = (umad_port_t *) 0x0
(gdb) p ca.ports[1]
$12 = (umad_port_t *) 0x6099d0
(gdb) bt
#0  resolve_ca_port (ca_name=ca_name@entry=0x611920 "mlx4_0", port=port@entry=0x7fffffffd82c) at ../libibumad/umad.c:292
#1  0x0000155554f0bea8 in resolve_ca_name (ca_in=0x0, ca_name=0x7fffffffd888, best_port=0x7fffffffd87c) at ../libibumad/umad.c:372
#2  resolve_ca_name (ca_in=<optimized out>, best_port=0x7fffffffd87c, ca_name=0x7fffffffd888) at ../libibumad/umad.c:334
#3  0x0000155554f0c22d in umad_open_port (ca_name=ca_name@entry=0x0, portnum=<optimized out>, portnum@entry=0) at ../libibumad/umad.c:701
#4  0x000015555511d718 in mad_rpc_open_port (dev_name=0x0, dev_port=0, mgmt_classes=mgmt_classes@entry=0x7fffffffdaa4, num_classes=num_classes@ent
ry=3) at ../libibmad/rpc.c:398
#5  0x00000000004019c4 in main (argc=3, argv=0x7fffffffdeb8) at ../infiniband-diags/ibportstate.c:423
> FWIW if a CA port is disabled it does not have a valid LID (unless running on an SM node). So I'm wondering why this is better than using a port GUID which should be unique.

Reply:
I think we should fix this code then.  The tool should be able to enable a local port.  I don't have time right now to figure out why this changed but at some point it did work and I don't see any reason it could not be made to work again.

Reply:
> I think we should fix this code then. 
Got it. I'm closing this PR and will open a new PR to fix the code.
Thanks.

Reply:
> Since when is enablement via GUID not working?

I tested the oldest OFED packages, which built from https://downloads.openfabrics.org/management/ , on a RHEL-7.0 x86-64 machine. They do not work.

```bash

~]$ ls *gz
infiniband-diags-1.3.2.tar.gz  libibcommon-1.0.5.tar.gz  libibumad-1.1.3.tar.gz
 libibmad-1.1.2.tar.gz     opensm-3.1.5.tar.gz    

 ~]$ rpm -q libibcommon libibumad opensm-libs libibmad infiniband-diags
libibcommon-1.0.5-1.el7.x86_64
libibumad-1.1.3-1.el7.x86_64
opensm-libs-3.1.5-1.el7.x86_64
libibmad-1.1.2-1.el7.x86_64
infiniband-diags-1.3.2-1.el7.x86_64

~]$ ibstat
CA 'mlx4_0'
	CA type: MT4099
	Number of ports: 2
	Firmware version: 2.42.5000
	Hardware version: 0
	Node GUID: 0xf4521403007be160
	System image GUID: 0xf4521403007be163
	Port 1:
		State: Active
		Physical state: LinkUp
		Rate: 56
		Base lid: 37
		LMC: 0
		SM lid: 13
		Capability mask: 0x02514868
		Port GUID: 0xf4521403007be161
	Port 2:
		State: Active
		Physical state: LinkUp
		Rate: 56
		Base lid: 2
		LMC: 0
		SM lid: 1
		Capability mask: 0x02514868
		Port GUID: 0xf4521403007be162

~]$ ibportstate 37 1 disable
ibportstate: iberror: failed: smp query nodeinfo: Node type not switch
```
I think `ibportstate` function `enable` needs the HCA name when multiple HCAs available. It also needs the port number for target HCA, which has more than one ports. Otherwise, the first active port of first HCA found by libibumad will be use. The first active port found by libibumad may not in the same fabric as the target port, which will be enabled.

When enablement via GUID for a system which has multiple HCAs or IB ports, the CA name and Port number was not specified, the first active port of first HCA found by libibumad will be used. The port selected by libibumad and the port specified by GUID may not in the same fabric. When they are not in the same fabric,  enablement via GUID will never work.

I think we need to update the help message to hint use to specific CA name and Port number when multiple HCAs or ports available. In the meanwhile, we also need to fix enablement via GUID when CA name and Port number are specified.